alpa/pipeline_parallel/stage_construction.py

```diff
@@ -82,7 +82,130 @@ def get_last_dp_result():
             last_autosharding_option_dicts)
 
 
-@maybe_numba_jit
+@maybe_numba_jit(fastmath=True, nogil=True, cache=True)
+def get_optimal_submeshes(best_s, f_argmin, num_devices, num_layers, submesh_n_devices):
+    current_s = best_s
+    current_layer = 0
+    current_devices = num_devices
+
+    res = []
+    while current_s > 0 and current_layer < num_layers and current_devices > 0:
+        next_start_layer, submesh_choice, autosharding_choice = (
+            f_argmin[current_s, current_layer, current_devices])
+        assert next_start_layer != -1 and current_devices != -1
+        res.append(((current_layer, next_start_layer), submesh_choice,
+                    autosharding_choice))
+        current_s -= 1
+        current_layer = next_start_layer
+        current_devices -= submesh_n_devices[submesh_choice]
+    assert current_s == 0 and current_layer == num_layers and current_devices == 0
+
+    return res
+
+
+@maybe_numba_jit(fastmath=True, nogil=True, cache=True)
+def dp_impl_2(
+    num_layers, num_devices, submesh_sizes, valid_idxs_and_costs, max_n_succ_stages
+):
+    f = np.full(
+        (num_layers + 1, num_layers + 1, num_devices + 1), np.inf, dtype=np.float32
+    )
+    f_stage_max = np.full(
+        (num_layers + 1, num_layers + 1, num_devices + 1), 0.0, dtype=np.float32
+    )
+    f_argmin = np.full(
+        (num_layers + 1, num_layers + 1, num_devices + 1, 3), -1, dtype=np.int32
+    )
+    f[0, num_layers, 0] = 0
+    for d in maybe_prange(1, num_devices + 1):
```

Comment:
There are dependencies between the iterations on different values of `d`. Can we actually use `prange` here? Even if the result is correct, can `prange` here actually accelerate the execution?